File: src/textIntegrity.ts
/**
 * Explicit text-boundary policy shared by bridge features.
 *
 * A JavaScript string is not automatically a safe UTF-8 value: it can contain
 * unpaired UTF-16 surrogates, which Node silently encodes as U+FFFD.  Keep the
 * check here, before a value is persisted, hashed, compared, or sent onward.
 *
 * This module intentionally does not offer a global `normalizeText()` helper.
 * Callers must state whether their field is human text, a derived search key,
 * verbatim content, or an opaque identifier.
 */

export type TextIntegrityErrorCode =
  | "TEXT_TYPE_INVALID"
  | "TEXT_UTF8_INVALID"
  | "TEXT_UNICODE_INVALID"
  | "TEXT_EMPTY"
  | "TEXT_TOO_LONG"
  | "TEXT_TOO_LARGE"
  | "TEXT_CONTROL_CHARACTER"
  | "TEXT_NUL";

export class TextIntegrityError extends Error {
  constructor(
    readonly code: TextIntegrityErrorCode,
    message: string,
    options?: ErrorOptions
  ) {
    super(message, options);
    this.name = "TextIntegrityError";
  }
}

export type TextIntegrityOptions = {
  /** Human-readable field name used only in a local validation error. */
  field?: string;
  /** Empty values are rejected unless explicitly allowed. */
  allowEmpty?: boolean;
  /** Bound Unicode scalar values, rather than UTF-16 code units. */
  maxCharacters?: number;
  /** Bound exact UTF-8 bytes after the selected policy is applied. */
  maxUtf8Bytes?: number;
  /** Reject C0/C1 control characters. */
  rejectControlCharacters?: boolean;
  /** Reject U+0000 independently of the broader control-character rule. */
  rejectNul?: boolean;
  /** Apply trim after NFC normalization. Only meaningful for human text. */
  trim?: boolean;
  /** Fold Unicode whitespace runs to one ASCII space after NFC normalization. */
  collapseWhitespace?: boolean;
};

const CONTROL_CHARACTERS = /[\u0000-\u001f\u007f-\u009f]/u;
const UNICODE_WHITESPACE = /\p{White_Space}+/gu;

/**
 * Decode an external byte boundary without Node's replacement-character
 * fallback.  The decoded text itself is also checked for well-formed UTF-16
 * before callers can serialize it again.
 */
export function decodeUtf8Strict(
  input: Uint8Array,
  field = "text"
): string {
  try {
    // `ignoreBOM: true` retains a BOM as content rather than silently removing
    // a byte sequence from verbatim values.
    const value = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(input);
    assertWellFormedUnicode(value, field);
    return value;
  } catch (error) {
    if (error instanceof TextIntegrityError) throw error;
    throw new TextIntegrityError(
      "TEXT_UTF8_INVALID",
      `${field} is not valid UTF-8.`,
      { cause: error }
    );
  }
}

/** Assert that a JavaScript string can round-trip as UTF-8 without mutation. */
export function assertWellFormedUnicode(value: unknown, field = "text"): asserts value is string {
  if (typeof value !== "string") {
    throw new TextIntegrityError("TEXT_TYPE_INVALID", `${field} must be text.`);
  }
  for (let index = 0; index < value.length; index += 1) {
    const codeUnit = value.charCodeAt(index);
    if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
      const following = value.charCodeAt(index + 1);
      if (following >= 0xdc00 && following <= 0xdfff) {
        index += 1;
        continue;
      }
      throw new TextIntegrityError(
        "TEXT_UNICODE_INVALID",
        `${field} contains an unpaired Unicode surrogate.`
      );
    }
    if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
      throw new TextIntegrityError(
        "TEXT_UNICODE_INVALID",
        `${field} contains an unpaired Unicode surrogate.`
      );
    }
  }
}

/**
 * Check every text value in an already-decoded JSON-shaped value. JSON can
 * spell an unpaired surrogate as `\\ud800`; byte-level UTF-8 validation alone
 * therefore is not enough after JSON.parse().
 */
export function assertJsonTextIntegrity(value: unknown, field = "JSON"): void {
  if (typeof value === "string") {
    assertWellFormedUnicode(value, field);
    return;
  }
  if (Array.isArray(value)) {
    value.forEach((entry, index) => assertJsonTextIntegrity(entry, `${field}[${index}]`));
    return;
  }
  if (value && typeof value === "object") {
    for (const [key, entry] of Object.entries(value)) {
      assertWellFormedUnicode(key, `${field} key`);
      assertJsonTextIntegrity(entry, `${field}.${key}`);
    }
  }
}

/**
 * Parse JSON received as bytes without allowing a lossy UTF-8 conversion or
 * an escaped unpaired surrogate to enter a bridge-owned record.
 */
export function parseJsonUtf8Strict<T = unknown>(
  input: Uint8Array,
  field = "JSON"
): T {
  const text = decodeUtf8Strict(input, field);
  return parseJsonTextStrict<T>(text, field);
}

/** Parse already-decoded text while retaining the escaped-surrogate check. */
export function parseJsonTextStrict<T = unknown>(
  input: string,
  field = "JSON"
): T {
  assertWellFormedUnicode(input, field);
  const value = JSON.parse(input) as unknown;
  assertJsonTextIntegrity(value, field);
  return value as T;
}

/** Exact byte length after verifying that UTF-8 encoding cannot replace data. */
export function utf8ByteLength(value: string, field = "text"): number {
  assertWellFormedUnicode(value, field);
  return Buffer.byteLength(value, "utf8");
}

/**
 * Canonical text for names, descriptions, tags, and other display/searchable
 * fields where canonically equivalent Unicode spellings mean the same thing.
 */
export function canonicalHumanText(value: unknown, options: TextIntegrityOptions = {}): string {
  const field = options.field || "text";
  assertWellFormedUnicode(value, field);
  let result = value.normalize("NFC");
  if (options.collapseWhitespace) result = result.replace(UNICODE_WHITESPACE, " ");
  if (options.trim) result = result.trim();
  assertTextConstraints(result, {
    ...options,
    field,
    allowEmpty: options.allowEmpty ?? false,
    rejectControlCharacters: options.rejectControlCharacters ?? true,
    rejectNul: options.rejectNul ?? true
  });
  return result;
}

/**
 * Verbatim content for prompts, Markdown, code, paths, command lines, hashes,
 * and opaque external values. This validates encoding and optional field
 * bounds but never normalizes, trims, folds whitespace, or changes line ends.
 */
export function verbatimText(value: unknown, options: TextIntegrityOptions = {}): string {
  const field = options.field || "text";
  assertWellFormedUnicode(value, field);
  assertTextConstraints(value, {
    ...options,
    field,
    allowEmpty: options.allowEmpty ?? false,
    rejectNul: options.rejectNul ?? true
  });
  return value;
}

/**
 * Opaque identifiers are validated for UTF-8 fidelity only. They intentionally
 * receive no Unicode or case normalization because they may be protocol IDs,
 * filesystem names, hashes, or credentials.
 */
export function opaqueIdentifier(value: unknown, options: TextIntegrityOptions = {}): string {
  return verbatimText(value, options);
}

/**
 * Derived comparison key.  NFC is enough for canonical-equivalence matching;
 * NFKC is deliberately not used because it can merge distinct identifiers.
 * This never changes the caller's stored presentation value.
 */
export function searchKey(value: unknown, options: TextIntegrityOptions = {}): string {
  const canonical = canonicalHumanText(value, {
    ...options,
    allowEmpty: options.allowEmpty ?? true,
    collapseWhitespace: options.collapseWhitespace ?? true,
    trim: options.trim ?? true
  });
  return canonical.toLocaleLowerCase("en-US").normalize("NFC");
}

function assertTextConstraints(value: string, options: TextIntegrityOptions): void {
  const field = options.field || "text";
  const allowEmpty = options.allowEmpty ?? false;
  if (!allowEmpty && value.length === 0) {
    throw new TextIntegrityError("TEXT_EMPTY", `${field} cannot be empty.`);
  }
  if (options.maxCharacters !== undefined && Array.from(value).length > options.maxCharacters) {
    throw new TextIntegrityError(
      "TEXT_TOO_LONG",
      `${field} exceeds ${options.maxCharacters} characters.`
    );
  }
  if (options.maxUtf8Bytes !== undefined && utf8ByteLength(value, field) > options.maxUtf8Bytes) {
    throw new TextIntegrityError(
      "TEXT_TOO_LARGE",
      `${field} exceeds ${options.maxUtf8Bytes} UTF-8 bytes.`
    );
  }
  if (options.rejectNul && value.includes("\u0000")) {
    throw new TextIntegrityError("TEXT_NUL", `${field} cannot contain NUL.`);
  }
  if (options.rejectControlCharacters && CONTROL_CHARACTERS.test(value)) {
    throw new TextIntegrityError(
      "TEXT_CONTROL_CHARACTER",
      `${field} cannot contain control characters.`
    );
  }
}

File: macos/Sources/CodexBridgeKit/TextIntegrity.swift
import Foundation

/// Field-scoped text policy shared with the Node bridge. Swift `String` is
/// already Unicode-scalar valid; byte boundaries still use strict UTF-8 and
/// callers select canonical, verbatim, opaque, or derived-search behavior.
public enum BridgeTextIntegrityError: Error, Equatable, Sendable {
    case invalidUTF8
    case invalidUnicode
    case empty
    case tooLong
    case tooLarge
    case controlCharacter
    case nul
}

public struct BridgeTextIntegrityOptions: Sendable {
    public var allowEmpty: Bool
    public var maxCharacters: Int?
    public var maxUTF8Bytes: Int?
    public var rejectControlCharacters: Bool
    public var rejectNul: Bool
    public var trim: Bool
    public var collapseWhitespace: Bool

    public init(
        allowEmpty: Bool = false,
        maxCharacters: Int? = nil,
        maxUTF8Bytes: Int? = nil,
        rejectControlCharacters: Bool = true,
        rejectNul: Bool = true,
        trim: Bool = false,
        collapseWhitespace: Bool = false
    ) {
        self.allowEmpty = allowEmpty
        self.maxCharacters = maxCharacters
        self.maxUTF8Bytes = maxUTF8Bytes
        self.rejectControlCharacters = rejectControlCharacters
        self.rejectNul = rejectNul
        self.trim = trim
        self.collapseWhitespace = collapseWhitespace
    }
}

public enum BridgeTextIntegrity {
    /// Reject malformed sequences rather than accepting Foundation's lossy
    /// replacement behavior at network, file, or local-socket byte boundaries.
    public static func decodeUTF8Strict(_ data: Data) throws -> String {
        // Foundation removes a leading UTF-8 BOM. Count and reinsert every
        // leading BOM so verbatim byte boundaries match Node and never discard
        // authored content before persistence, hashing, or transport.
        let bom = Data([0xef, 0xbb, 0xbf])
        var prefixBytes = 0
        while data.dropFirst(prefixBytes).starts(with: bom) {
            prefixBytes += bom.count
        }
        guard let suffix = String(data: data.dropFirst(prefixBytes), encoding: .utf8) else {
            throw BridgeTextIntegrityError.invalidUTF8
        }
        return String(repeating: "\u{feff}", count: prefixBytes / bom.count) + suffix
    }

    public static func canonicalHumanText(
        _ value: String,
        options: BridgeTextIntegrityOptions = .init()
    ) throws -> String {
        var result = value.precomposedStringWithCanonicalMapping
        if options.collapseWhitespace {
            result = result.split(whereSeparator: { $0.isWhitespace }).joined(separator: " ")
        }
        if options.trim {
            result = result.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        try validate(result, options: options)
        return result
    }

    /// Never applies NFC, trimming, or line-ending conversion. Use this for
    /// prompts, Markdown, paths, commands, identifiers, and secrets.
    public static func verbatimText(
        _ value: String,
        options: BridgeTextIntegrityOptions = .init(rejectControlCharacters: false)
    ) throws -> String {
        try validate(value, options: options)
        return value
    }

    public static func opaqueIdentifier(
        _ value: String,
        options: BridgeTextIntegrityOptions = .init(rejectControlCharacters: false)
    ) throws -> String {
        try verbatimText(value, options: options)
    }

    /// A derived comparison key only; this must never be stored in place of
    /// the presentation value. NFKC is intentionally not used.
    public static func searchKey(
        _ value: String,
        options: BridgeTextIntegrityOptions = .init(
            allowEmpty: true,
            trim: true,
            collapseWhitespace: true
        )
    ) throws -> String {
        let canonical = try canonicalHumanText(
            value,
            options: options
        )
        return canonical.lowercased(with: Locale(identifier: "en_US"))
            .precomposedStringWithCanonicalMapping
    }

    /// Validate byte encoding before Foundation JSON decoding. JSON is allowed
    /// to carry only valid UTF-8 and scalar-value strings on this transport.
    /// Foundation can otherwise accept an escaped unpaired UTF-16 surrogate
    /// that Node's JSON parser would expose as a lossy JavaScript string.
    public static func validateJSONUTF8(_ data: Data) throws {
        let text = try decodeUTF8Strict(data)
        // JSON.parse rejects a leading BOM. Foundation's JSONDecoder accepts
        // one, so reject it here to keep both bridge implementations aligned.
        if text.first == "\u{feff}" { throw BridgeTextIntegrityError.invalidUnicode }
        try validateJSONUnicodeEscapes(Array(data))
    }

    private static func validate(_ value: String, options: BridgeTextIntegrityOptions) throws {
        if !options.allowEmpty && value.isEmpty { throw BridgeTextIntegrityError.empty }
        // Match Node's Array.from(value).length: count Unicode scalar values,
        // not Swift grapheme clusters. A decomposed character therefore uses
        // the same field budget on both sides of the RPC boundary.
        if let maximum = options.maxCharacters, value.unicodeScalars.count > maximum {
            throw BridgeTextIntegrityError.tooLong
        }
        if let maximum = options.maxUTF8Bytes, value.lengthOfBytes(using: .utf8) > maximum {
            throw BridgeTextIntegrityError.tooLarge
        }
        if options.rejectNul && value.unicodeScalars.contains("\u{0000}") {
            throw BridgeTextIntegrityError.nul
        }
        if options.rejectControlCharacters && value.unicodeScalars.contains(where: {
            $0.properties.generalCategory == .control
        }) {
            throw BridgeTextIntegrityError.controlCharacter
        }
    }

    private static func validateJSONUnicodeEscapes(_ bytes: [UInt8]) throws {
        var index = 0
        var inString = false
        while index < bytes.count {
            let byte = bytes[index]
            if !inString {
                if byte == 0x22 { inString = true }
                index += 1
                continue
            }
            if byte == 0x22 {
                inString = false
                index += 1
                continue
            }
            guard byte == 0x5c else {
                index += 1
                continue
            }
            guard index + 1 < bytes.count else { throw BridgeTextIntegrityError.invalidUnicode }
            guard bytes[index + 1] == 0x75 else {
                index += 2
                continue
            }
            guard let first = unicodeEscapeCodeUnit(bytes, at: index + 2) else {
                throw BridgeTextIntegrityError.invalidUnicode
            }
            if (0xd800...0xdbff).contains(first) {
                guard index + 11 < bytes.count,
                      bytes[index + 6] == 0x5c,
                      bytes[index + 7] == 0x75,
                      let second = unicodeEscapeCodeUnit(bytes, at: index + 8),
                      (0xdc00...0xdfff).contains(second) else {
                    throw BridgeTextIntegrityError.invalidUnicode
                }
                index += 12
                continue
            }
            if (0xdc00...0xdfff).contains(first) {
                throw BridgeTextIntegrityError.invalidUnicode
            }
            index += 6
        }
    }

    private static func unicodeEscapeCodeUnit(_ bytes: [UInt8], at index: Int) -> UInt16? {
        guard index + 3 < bytes.count else { return nil }
        var value: UInt16 = 0
        for offset in 0..<4 {
            guard let digit = hexadecimalValue(bytes[index + offset]) else { return nil }
            value = (value << 4) | digit
        }
        return value
    }

    private static func hexadecimalValue(_ byte: UInt8) -> UInt16? {
        switch byte {
        case 0x30...0x39: return UInt16(byte - 0x30)
        case 0x41...0x46: return UInt16(byte - 0x41 + 10)
        case 0x61...0x66: return UInt16(byte - 0x61 + 10)
        default: return nil
        }
    }
}
